### ISSUE TITLE
"`*const u8` cannot be shared between threads safely"

### ISSUE THREAD

**Issue by JoelSatkas** [google/autocxx#1336]
**Describe the bug**
Cannot pass rust created c++ objects to other threads.

**To Reproduce**

C++ code
```
class CCPClass
{
public:
	void Initialize() 
        {
            std::cout << "Function called from c++" << std::endl;
        }
};
```

Rust
```
pub use ffi::*;
use std::{sync::Arc, thread, pin::Pin};

include_cpp! {
  #include "CCPClass.h"
  safety!(unsafe)
  generate!("CCPClass")
}

#[derive(Clone)]
pub struct CppInterfaces {
    cpp_utility1_interface: Arc<Pin<Box<CCPClass>>>,
}

impl CppInterfaces {
    fn new() -> Self {
        let cpp_utility1_interface= Arc::new(CCPClass::new().within_box());
        CppInterfaces { cpp_utility1_interface}
    }
}

fn main () -> {
  let cpp_object = Arc::new(CCPClass::new());
  let handle = thread::spawn(|| {
        other_mod::do_thing_with_cpp_interface(cpp_object); // <-- error is here
  });

...
}

mod other_mod {
fn do_thing_with_cpp_interface(cpp_interfaces: Arc<CppInterfaces>) {
...
}
}
```

Error message:
```
`*const u8` cannot be shared between threads safely
within `root::CCPClass`, the trait `Sync` is not implemented for `*const u8`
required because it appears within the type `[*const u8; 0]`
required for `Unique<root::CCPClass>` to implement `Sync`
required for `Arc<Pin<Box<root::CCPClass>>>` to implement `Sync`
required for `Arc<CppInterfaces>` to implement `std::marker::Send`rustc

my_rust_file.rs(83, 22): required by a bound introduced by this call
marker.rs(791, 12): required because it appears within the type `PhantomData<[*const u8; 0]>`
autocxx-ffi-default-gen.rs(1, 9758): required because it appears within the type `CCPClass`
boxed.rs(195, 12): required because it appears within the type `Box<CCPClass>`
pin.rs(410, 12): required because it appears within the type `Pin<Box<CCPClass>>`
my_rust_file.rs(35, 12): required because it appears within the type `CppInterfaces`
my_rust_file.rs(83, 36): required because it's used within this closure
mod.rs(683, 8): required by a bound in `std::thread::spawn`
````

Basically, after creating a struct that wraps an Arc to the autocxx created c++ object and trying to pass this struct to another thread i get this error that a raw pointer is not safe to pass between threads compleltly preventing me from implementing my project which uses multiple threads from a tokio pool to access the cpp object in parallel. In my case, the c++ object itself has mutex protection around data and takes care of all of that.

Looking at the autocxx generated code i can see this:

```
#[doc = " <summary>\n \n </summary>"]
#[repr(C, align(1))]
pub struct CCPClass {
    _pinned: core::marker::PhantomData<core::marker::PhantomPinned>,
    _non_send_sync: core::marker::PhantomData<[*const u8; 0]>,
    _data: [u8; 1],
}
```

I can see that the `_non_send_sync` is the thing preventing the copy to happen, what is this variable and what is it used for? Maybe its possible that i do not need this in my use case or to make this sync compatible. 

**Expected behavior**
Would love to pass c++ objects between threads.

**Additional context**
If it helps, i can mention that I want the c++ object to be given as state to axum which will use the c++ object on user requests to pass the information into the c++ runtime and return. Right now this issue is preventing me from doing that. BTW im a rust noob and recently started to learn it so apologies if this is a stupid or irrelevant question/issue, ive definitely asked plenty of those before as i slowly learn rust.

**Comment by adetaylor** [comment]
I think the workaround is [described here](https://cxx.rs/extern-c++.html?highlight=sync#opaque-c-types). Let me know if this works. If it does, I should probably add this to some autocxx documentation somewhere, so I'm not going to close this just yet.

**Comment by JoelSatkas** [comment]
> I think the workaround is [described here](https://cxx.rs/extern-c++.html?highlight=sync#opaque-c-types). Let me know if this works. If it does, I should probably add this to some autocxx documentation somewhere, so I'm not going to close this just yet.

Ok, i presumed you meant the **Thread safety** section and added the sync and send traits for my c++ class like:

```
/// The C++ implementation of MyType is thread safe.
unsafe impl Send for CCPClass {

}
unsafe impl Sync for CCPClass {

}
```

And now it no longer complains and works. This is great, thanks! It also works during run time, i can only presume the sync and send traits are like tags and are supposed to be left empty because i dont have any code there and it works. I would definitely recommend updating the documentation, i found it a bit lacking for multi threading information. 

I now have an issue with autocxx not being able to generate c++ functions taking in a vector of unsigned ints but i presume i should make a new issue for that. 

Thanks again.

**Comment by JoelSatkas** [comment]
> I now have an issue with autocxx not being able to generate c++ functions taking in a vector of unsigned ints but i presume i should make a new issue for that.

Saw other issues, changed unsigned int to uint32_t and now it works also so nvm
